### ISSUE TITLE
Feature Request: Fragments Components (like ng-container)

### ISSUE THREAD

**Issue by michaeljota** [angular/angular#22208]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[ ] Bug report  
[x] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

Every custom component that you can make with Angular are nodes in the document as a new tag.

## Expected behavior

I would like a way to create a custom component that does not show it self in the document tree, but only it's content. 

Just like `ng-container` tag does. 
## What is the motivation / use case for changing the behavior?

I'm using Angular Material to the application UI, and some of their components needs to be child of each other. However, the content of their children could be separated if there was a way to create a component that does not add a middle element in the document tree. 

## Environment

<pre><code>
Angular version: 5.2.3

Browser:
- [ ] Chrome (desktop) version XX
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: XX  
- Platform:  

Others:

</code></pre>
Related:
- Issues
  - #7857
- Other SPA Platforms:
  - [React Fragments](https://reactjs.org/docs/react-component.html#fragments)

**Comment by IgorMinar** [comment]
This is already possible via lower level apis - a component can define a template with embedded templates and those can then be used be instantiated and projected into any view container.

What we are missing is a declarative way to define these fragments without the need for involving all of the low level apis.

**Comment by michaeljota** [comment]
I update the issue with suggestions about how this could be. I don't know much about the Angular low level APIs, but I remember this behavior was dropped from AnguarJS, or a least something very similar.

**Comment by trotyl** [comment]
## Things to note:

### Style cannot be native anymore

Of course.

### Style cannot be emulated anymore

```typescript
@Component({
  selector: 'fragment1',
  template: `
    <p>Hello World!</p>
  `,
  styles: [`p { color: red; }`],
})
class Fragment1 {}

@Component({
  selector: 'fragment1',
  template: `
    <fragment1></fragment1>
  `,
  styles: [`p { color: blue; }`],
})
class Fragment2 {}

@Component({
  selector: 'my-comp',
  template: `
    <fragment2></fragment2>
  `,
  styles: [`p { color: green; }`],
})
class MyComponent {}
```

Will result to:

```html
<style>p[_ngcontent-c0] { color: green; }</style>
<style>p[_ngcontent-c1] { color: blue; }</style>
<style>p[_ngcontent-c2] { color: red; }</style>

<my-comp _nghost-c0>
  <p _ngcontent-c0 _ngcontent-c1 _ngcontent-c2>Hello World!</p>
</my-comp>
```

The final style would only be determined by insertion order of `<style>` tags, which can never be relied on.

## Technically a fragment could **NOT** have:

+ `styleUrls`/`style`;
+ `encapsulation`;
+ `entryComponents`; (Why is it even in `Component`?)
+ `host`;

So it may not be a `Component` or `Directive`, but under its own metadata. (eg. `@Fragment()`)

**Comment by michaeljota** [comment]
> just not possible for EmbeddedView to have style.

So computed encapsulated styles could not me applied to the children? Like in your example

**Comment by trotyl** [comment]
@michaeljota Not like my example, that has been commented already. Styles are all [static to `Renderer`](https://angular.io/api/core/RendererType2), just not possible to achieve that in current architecture. There could not even be a "what if".

**Comment by michaeljota** [comment]
I agree that the issue is the same, but I open this not about the issue, but a proposal to solve it, and discussion around it.

**Comment by pkozlowski-opensource** [comment]
> I agree that the issue is the same, but I open this not about the issue, but a proposal to solve it, and discussion around it.

Sure. We can still centralize the discussion in one place. Would you mind coping your suggestions over to #18877? I'm digging into this topic and I would like to have all the info in one place.

**Comment by michaeljota** [comment]
@pkozlowski-opensource I commented the proposal in the original post. I update it with more info about the usage, and the declarations. What do you think about it? I would be glad to help in the development of this.
